kate/recovery/src/data.rs

```diff
@@ -42,17 +45,248 @@ impl Cell {
+				if i >= Self::LIMBS_PER_SCALAR {
+					return Err("Too many limbs in scalar");
+				}
+				scalar[i] = limb_bytes
+					.try_into()
+					.ok()
+					.map(u64::from_be_bytes)
+					.ok_or("Failed to decode scalar limb")?;
+			}
+			scalars.push(scalar);
+		}
+
+		Ok(Self {
+			position,
+			proof,
+			gcell_block,
+			scalars,
+		})
+	}
+
+	pub fn to_bytes(&self) -> Vec<u8> {
+		let proof_bytes: Vec<u8> = self.proof.into();
+		let gcell_block_bytes = self.gcell_block.to_bytes();
+		let scalar_count = self.scalars.len();
+
+		let total_size = Self::PROOF_BYTE_LEN
+			+ gcell_block_bytes.len()
+			+ Self::SCALAR_COUNT_LEN
+			+ scalar_count * Self::SCALAR_BYTE_LEN;
+
+		let mut content = Vec::with_capacity(total_size);
+		content.extend_from_slice(&proof_bytes);
+		content.extend_from_slice(&gcell_block_bytes);
+		content.extend_from_slice(&(scalar_count as u32).to_le_bytes());
+		content.extend_from_slice(&self.data());
+
+		content
+	}
+
+	pub fn data(&self) -> Vec<u8> {
+		let mut bytes = Vec::with_capacity(self.scalars.len() * Self::BYTES_PER_SCALAR);
+
+		for scalar in &self.scalars {
+			for &limb in scalar {
+				bytes.extend_from_slice(&limb.to_be_bytes());
+			}
+		}
+
+		bytes
+	}
+
+	pub fn proof(&self) -> [u8; 48] {
+		self.proof
+	}
+}
+
+#[derive(Debug, Clone)]
+pub enum Cell {
+	SingleCell(SingleCell),
+	MultiProofCell(MultiProofCell),
+}
+
+impl Cell {
+	#[cfg(any(target_arch = "wasm32", feature = "std"))]
+	pub fn reference(&self, block: u32) -> String {
+		match self {
+			Cell::SingleCell(cell) => cell.reference(block),
+			Cell::MultiProofCell(mcell) => mcell.reference(block),
+		}
+	}
+
+	pub fn data(&self) -> Vec<u8> {
+		match self {
+			Cell::SingleCell(cell) => cell.data().to_vec(),
+			Cell::MultiProofCell(mcell) => mcell.data(),
+		}
+	}
+
+	pub fn position(&self) -> Position {
+		match self {
+			Cell::SingleCell(cell) => cell.position,
+			Cell::MultiProofCell(mcell) => mcell.position,
+		}
+	}
+
+	pub fn proof(&self) -> [u8; 48] {
+		match self {
+			Cell::SingleCell(cell) => cell.proof(),
+			Cell::MultiProofCell(mcell) => mcell.proof(),
+		}
+	}
+
+	pub fn to_bytes(&self) -> Vec<u8> {
+		match self {
+			Cell::MultiProofCell(mcell) => mcell.to_bytes(),
+			Cell::SingleCell(cell) => cell.data().to_vec(),
+		}
+	}
+}
+
+impl From<SingleCell> for Cell {
+	fn from(cell: SingleCell) -> Self {
+		Cell::SingleCell(cell)
+	}
+}
+
+impl From<MultiProofCell> for Cell {
+	fn from(mcell: MultiProofCell) -> Self {
+		Cell::MultiProofCell(mcell)
+	}
+}
+
+impl TryFrom<Cell> for SingleCell {
+	type Error = &'static str;
+
+	fn try_from(value: Cell) -> Result<Self, Self::Error> {
+		match value {
+			Cell::SingleCell(cell) => Ok(cell),
+			Cell::MultiProofCell(_) => Err("Expected SingleCell, found MultiProofCell"),
+		}
+	}
+}
+
 /// Merges cells data per row.
 /// Cells are sorted before merge.
 pub fn rows(dimensions: Dimensions, cells: &[&Cell]) -> Vec<(RowIndex, Vec<u8>)> {
```

Comment:
I think we need to re-examine every function which takes a Cell as param as we've changed the Cell's definition, eg: in this function we're filtering based on `row_byte_size()` but that function corresponds to only SingleCell kind with new defn & if we pass `MultiProofCell`, this can lead to data_loss. If this function is intended to take only `SingleCell` type, disallow multiproof cell type.